python/singa/autograd.py

```diff
@@ -3330,6 +3330,94 @@ def step_forward(self, x, h, c, Wx, Wh, Bx, Bh):
         return hout, cout
 
 
+class _RNN(Operation):
+    """ RNN operation with c++ backend
+    """
+    def __init__(self, handle):
+        assert singa.USE_CUDA is True, "Not able to run without CUDA"
+        super(_RNN, self).__init__()
+        self.handle = handle
+
+    def forward(self, x, W):
+        # TODO: CPU forward
+
+        # GPU forward
+        if training:
+            y = singa.GpuRNNForwardTraining(x, W, self.handle)
+            self.inputs = (x, W, y)
+        else:
+            y = singa.GpuRNNForwardInference(x, W, self.handle)
+
+        return y
+
+    def backward(self, dy):
+        assert training is True and hasattr(
+            self, "inputs"), "Please set training as True before do BP. "
+
+        # TODO: CPU backward
+
+        # GPU backward
+        dx = singa.GpuRNNBackwardx(self.inputs[2], dy, self.inputs[1], self.handle)
+        dW = singa.GpuRNNBackwardW(self.inputs[0], self.inputs[2], self.handle)
+        return dx, dW
+
+class RNN_direct(Layer):
```

Comment:
rename to RNN?

Reply:
RNN is used for existing highlevel implementation
After consideration, RNN_direct is used to name newly added C++ backed (cudnn/cpu) RNN

Reply:
let's unify all the rnn implementation in this way

```python
def RNN(hidden_size, num_layers=1, dropout=0.0, direction='one', return_seq=False, return_state=False, alg='lstm', backend='cudnn') :
   if backend == 'cudnn':
      reutrn CudnnRNN(hidden_size, num_layers, dropout, direction, mode)
   if backend == 'native':
     if mode == 'lstm':
      return LSTM()
     elif mode == 'vanilla':
       return VanillaRNN()
     ....
```

All rnn classes inherit 
```python
class RNNBase(Layer):
   def __init__(self, hidden_size, num_layers=1, dropout=0.0, direction='one', return_seq=False, return_state=False, alg='lstm')
```

Reply:
this is rectified